django/db/backends/postgresql/base.py

```diff
@@ -212,7 +260,17 @@ def get_connection_params(self):
 
     @async_unsafe
     def get_new_connection(self, conn_params):
-        connection = Database.connect(**conn_params)
+        if self.is_psycopg3:
+            ctx = get_adapters_template(settings.USE_TZ, self.timezone)
+            connection = self.Database.connect(**conn_params, context=ctx)
+        else:
+            connection = self.Database.connect(**conn_params)
+            # Register dummy loads() to avoid a round trip from psycopg2's decode
+            # to json.dumps() to json.loads(), when using a custom decoder in
+            # JSONField.
+            psycopg2.extras.register_default_jsonb(
+                conn_or_curs=connection, loads=lambda x: x
+            )
```

Comment:
What I mean by that is something along this lines (to avoid the branching at runtime)

```python
if PSYCOPG_VERSION >= (3, 0, 0):
    def _get_new_connection(self, conn_params):
        ctx = get_adapters_template(settings.USE_TZ, self.timezone)
        return self.Database.connect(**conn_params, context=ctx)
else:
    def _get_new_connection(self, conn_params):
        connection = self.Database.connect(**conn_params)
        # Register dummy loads() to avoid a round trip from psycopg2's decode
        # to json.dumps() to json.loads(), when using a custom decoder in
        # JSONField.
        psycopg2.extras.register_default_jsonb(
            conn_or_curs=connection, loads=lambda x: x
        )

And then use `_get_new_connection` in `get_new_connection`